src/battle_game/core/units/random_tank.cpp

```diff
@@ -0,0 +1,124 @@
+    : Tank(game_core, id, player_id) {
+}
+
+void RandomTank::Render() {
+  Tank::Render();
+}
+
+void RandomTank::TankMove(float move_speed, float rotate_angular_speed) {
+  auto player = game_core_->GetPlayer(player_id_);
+  if (player) {
+    auto &input_data = player->GetInputData();
+    glm::vec2 offset{0.0f};
+    bool fx = 0;
+    if (input_data.key_down[GLFW_KEY_W]) {
+      offset.y += 1.0f;
+      fx = 1;
+    }
+    if (input_data.key_down[GLFW_KEY_S]) {
+      offset.y -= 1.0f;
+      fx = 1;
+    }
+    float speed = move_speed * GetSpeedScale();
+    offset *= kSecondPerTick * speed;
+    auto new_position =
+        position_ + glm::vec2{glm::rotate(glm::mat4{1.0f}, rotation_,
+                                          glm::vec3{0.0f, 0.0f, 1.0f}) *
+                              glm::vec4{offset, 0.0f, 0.0f}};
+    if (fx) {
+      glm::vec2 offset2{0.0f};
+      offset2.y += (game_core_->RandomFloat() - 0.5);
+      offset2 *= kSecondPerTick * speed;
+      new_position = new_position +
+                     glm::vec2{glm::rotate(glm::mat4{1.0f},
+                                           (float)(rotation_ + acos(-1) / 2),
+                                           glm::vec3{0.0f, 0.0f, 1.0f}) *
+                               glm::vec4{offset2, 0.0f, 0.0f}};
+    }
+    if (!game_core_->IsBlockedByObstacles(new_position)) {
+      game_core_->PushEventMoveUnit(id_, new_position);
+    } else if (!game_core_->IsOutOfRange(new_position)) {
+      auto x = game_core_->GetBlockedObstacle(new_position);
+      if (x->GetObstacleType() == mud) {
+        new_position =
+            position_ + glm::vec2{glm::rotate(glm::mat4{1.0f}, rotation_,
+                                              glm::vec3{0.0f, 0.0f, 1.0f}) *
+                                  glm::vec4{offset / 3.0f, 0.0f, 0.0f}};
+
+        if (fx) {
+          glm::vec2 offset2{0.0f};
+          offset2.y += (game_core_->RandomFloat() - 0.5);
+          offset2 *= kSecondPerTick * speed;
+          new_position =
+              new_position +
+              glm::vec2{glm::rotate(glm::mat4{1.0f},
+                                    (float)(rotation_ + acos(-1) / 2),
+                                    glm::vec3{0.0f, 0.0f, 1.0f}) *
+                        glm::vec4{offset2 / 3.0f, 0.0f, 0.0f}};
+        }
+        if (!game_core_->IsBlockedByObstacles(new_position) ||
+            game_core_->GetBlockedObstacle(new_position)->GetObstacleType() ==
+                mud) {
+          game_core_->PushEventMoveUnit(id_, new_position);
+        }
+      }
+    }
+    float rotation_offset = 0.0f;
+    if (input_data.key_down[GLFW_KEY_A]) {
+      rotation_offset += 1.0f;
+    }
+    if (input_data.key_down[GLFW_KEY_D]) {
+      rotation_offset -= 1.0f;
+    }
+    rotation_offset *= kSecondPerTick * rotate_angular_speed * GetSpeedScale();
+    game_core_->PushEventRotateUnit(id_, rotation_ + rotation_offset);
+  }
+}
```

Comment:
Did you make any change? if not, inherit from unit::Tank too.

Reply:
I added the speed reduction of mud and random move in this function